helpers/basic_auth/LDAP/squid_ldap_auth.8
.TH squid_ldap_auth 8 "17 July 2004" "Squid LDAP Auth"
.
.SH NAME
squid_ldap_auth - Squid LDAP authentication helper
.
.SH SYNOPSIS
.B squid_ldap_auth
-b "base DN" [-u attribute] [options] [ldap_server_name[:port]|URI]...
.br
.B squid_ldap_auth
-b "base DN" -f "LDAP search filter" [options] [ldap_server_name[:port]|URI]...
.
.SH DESCRIPTION
This helper allows Squid to connect to a LDAP directory to
validate the user name and password of Basic HTTP authentication.
.P
The program has two major modes of operation. In the default mode
of operation the users DN is constructed using the base DN and
user attribute. In the other mode of operation a search
filter is used to locate valid user DN's below the base DN.
.
.TP
.BI "-b " "basedn " (REQUIRED)
Specifies the base DN under which the users are located.
.
.TP
.BI "-f " filter
LDAP search filter to locate the user DN. Required if the users
are in a hierarchy below the base DN, or if the login name is
not what builds the user specific part of the users DN.
.IP
The search filter can contain up to 15 occurrences of %s
which will be replaced by the username, as in "uid=%s" for
RFC2037 directories. For a detailed description of LDAP search
filter syntax see RFC2254.
.
.TP
.BI "-u " userattr
Specifies the name of the DN attribute that contains the username/login.
Combined with the base DN to construct the users DN when no search filter
is specified (-f option). Defaults to 'uid'
.IP
Note: This can only be done if all your users are located directly under
the same position in the LDAP tree and the login name is used for naming
each user object. If your LDAP tree does not match these criterias or if
you want to filter who are valid users then you need to use a search filter
to search for your users DN (-f option).
.
.TP
.BI "-U " passwordattr
Use ldap_compare instead of ldap_simple_bind to verify the users password.
passwordattr is the LDAP attribute storing the users password.
.
.TP
.BI "-s " base|one|sub
search scope when performing user DN searches specified
by the -f option. Defaults to 'sub'.
.IP
.B base
object only,
.B one
level below the base object or
.BR sub tree
below the base object
.
.TP
.BI "-D " "binddn " "-w " password
The DN and password to bind as while performing searches. Required by the
.BI -f
flag if the directory does not allow anonymous searches.
.IP
As the password needs to be printed in plain text in your Squid configuration
it is strongly recommended to use a account with minimal associated privileges.
This to limit the damage in case someone could get hold of a copy of your
Squid configuration file.
.
.TP
.BI "-D " "binddn " "-W " "secretfile "
The DN and the name of a file containing the password
to bind as while performing searches. 
.IP
Less insecure version of the former parameter pair with two advantages:
The password does not occur in the process listing, 
and the password is not being compromised if someone gets the squid 
configuration file without getting the secretfile.
.
.TP
.BI -P
Use a persistent LDAP connection. Normally the LDAP connection
is only open while validating a username to preserve resources
at the LDAP server. This option causes the LDAP connection to
be kept open, allowing it to be reused for further user
validations. Recommended for larger installations.
.
.TP
.BI -O
Only bind once per LDAP connection. Some LDAP servers do not
allow re-binding as another user after a successful ldap_bind.
The use of this option always opens a new connection for each
login attempt. If combined with the -P option for persistent
LDAP connection then the connection used for searching for the
user DN is kept persistent but a new connection is opened
to verify each users password once the DN is found.
.
.TP
.BI -R
do not follow referrals
.
.TP
.BI "-a " never|always|search|find
when to dereference aliases. Defaults to 'never'
.IP
.BI never
dereference aliases (default),
.BI always
dereference aliases, only while
.BR search ing
or only to
.B find
the base object
.
.TP
.BI -H " ldapuri"
Specity the LDAP server to connect to by LDAP URI (requires OpenLDAP libraries).
Servers can also be specified last on the command line.
.
.TP
.BI -h " ldapserver"
Specify the LDAP server to connect to. Servers can also be specified last
on the command line.
.TP
.BI -p " ldapport"
Specify an alternate TCP port where the ldap server is listening if
other than the default LDAP port 389. Can also be specified within the
server specificiation by using servername:port syntax.
.
.TP
.BI -v " 2|3"
LDAP protocol version. Defaults to 2 if not specified.
.
.TP
.BI -Z
Use TLS encryption
.
.TP
.BI -S certpath
Enable LDAP over SSL (requires Netscape LDAP API libraries)
.
.TP
.BI -c connect_timeout
Specify timeout used when connecting to LDAP servers (requires
Netscape LDAP API libraries)
.TP
.BI -t search_timeout
Specify time limit on LDAP search operations
.
.TP
.BI -d
Debug mode where each step taken will get reported in detail.
Useful for understanding what goes wrong if the results is
not what is expected.
.
.SH EXAMPLES
For directories using the RFC2307 layout with a single domain, all
you need to specify is usually the base DN under where your users
are located and the server name:
.IP
squid_ldap_auth -b "ou=people,dc=your,dc=domain" ldapserver
.P
If you have sub-domains then you need to use a search filter approach
to locate your user DNs as these can no longer be constructed direcly
from the base DN and login name alone:
.IP
squid_ldap_auth -b "dc=your,dc=domain" -f "uid=%s" ldapserver
.P
And similarily if you only want to allow access to users having a
specific attribute
.IP
squid_ldap_auth -b "dc=your,dc=domain" -f "(&(uid=%s)(specialattribute=value))" ldapserver
.P
Or if the user attribute of the user DN is "cn" instead of "uid" and
you do not want to have to search for the users then you could use something
like the following example for Active Directory:
.IP
squid_ldap_auth -u cn -b "cn=Users,dc=your,dc=domain" ldapserver
.P
If you want to search for the user DN and your directory does not allow
anonymous searches then you must also use the -D and -w flags to specify
a user DN and password to log in as to perform the searches, as in the
following complex Active Directory example
.IP
squid_ldap_auth -p -R -b "dc=your,dc=domain" -D "cn=squid,cn=users,dc=your,dc=domain" -w "secretsquidpassword" -f "(&(userPrincipalName=%s)(objectClass=Person))" activedirectoryserver
.
.SH NOTES
.
When constructing search filters it is strongly recommended to test the filter
using ldapsearch before you attempt to use squid_ldap_auth. This to verify
that the filter matches what you expect.
.
.SH AUTHOR
This manual page was written by 
.I Henrik Nordstrom <hno@squid-cache.org>
.P
squid_ldap_auth is written by 
.I Glenn Newton <gnewton@wapiti.cisti.nrc.ca>
and
.I Henrik Nordstrom <hno@squid-cache.org>
.
.SH KNOWN ISSUES
Will crash if other % values than %s is used in -f, or if more than 15 %s
is used.
.
.SH QUESTIONS
Any questions on usage can be sent to 
.IR "Squid Users <squid-users@squid-cache.org>" ,
or to your favorite LDAP list/friend if the question is more related to
LDAP than Squid.
.
.SH REPORTING BUGS
Report bugs or bug-fixes to
.I Squid Bugs <squid-bugs@squid-cache.org>
or ideas for new improvements to 
.I Squid Developers <squid-dev@squid-cache.org>
.
.SH "SEE ALSO"
.BR ldapsearch ( 1 ),
.br
Your favorite LDAP documentation
.br
.BR RFC2254 " - The String Representation of LDAP Search Filters,"


ChangeLog
Changes to squid-2.6.STABLE24 (13 February 2010)

	- Bug #2365: cachemgr.cgi fails to HTML encode config dumps properly
	- Bug #2458: reply_body_max_size incorrectly documented
	- Bug #2858: Segment violation in HTCP
	- Improve %nn parser to better deal with certain odd %nn sequences
	- Segmentation fault if failed to open cache.log
	- Handle DNS header-only packets as invalid. (CVE-2010-0308)
	- Cosmetic change to keep GCC happy
	- Bug #2773: Segfault in RFC2069 Digest authantication
	- Bug #2819: const correctness errors in dns_internal.c

Changes to squid-2.6.STABLE23 (17 September 2009)
	- Bug #2494: Fix tproxy url in configure
	- Add in some better documentation for override-expire.
	- Fix Bug #2661 - Solaris /dev/poll support broken with EINVAL
	- Clarify external_acl_type %{Header} sightly
	- Cosmetic change to make GCC happy
	- Windows port: Fix PSAPI.DLL usage, is always available on Windows NT
	  and later
	- Windows port: Added support for Windows 7, Windows Server 2008 R2
	  and later
	- The debug mode option '-d' was not documented in LDAP helpers usage
	  message
	- Windows port: Added a note about installation on Windows Vista and
	  later
	- Bug #2642: Remove duplicate peerMonitorInit() on reconfigure
	- Bug #2672: cacheMemMaxSize 32-bit overflow during snmpwalk
	- Accept PUT/POST requests without an entity-body
	- Bug #2369: squid traffic counter 32-bit overflow
	- Bug #2080: wbinfo_group.pl - false positive under certain conditions
	- Bug #2739: DNS resolver option ndots can't be parsed from
	  resolv.conf
	- Bug #2541: Hang in 100% CPU loop while extacting header details
	  using a delimiter other than comma (external_acl_type,
	  access_log_format, external_refresh_check)
	- Bug #2768 - squid_ldap_group argument parsing error

Changes to squid-2.6.STABLE22 (19 October 2008)
	- Bug #2396: Correct the opening of the PF device file.
	- Make --with-large-files and --with-build-envirnment=default play
	  nice together
	- Workaround for Linux-2.6.24 & 2.6.25 netfiler_ipv4.h include header
	  __u32 problem
	- Make dns_nameserver work when using --disable-internal-dns on glibc
	  based systems
	- Bug #2426: Increase negotiate auth token buffer size
	- Bug #2427: squid_ldap_group -h reports the old % codes for -f
	- Bug #2477: swap.state permission issues if crashing during "squid -k
	  reconfigure"
	- Windows port: Fix build error using latest MinGW runtime.

Changes to squid-2.6.STABLE21 (27 June 2008)

	- Bug #2350: Bugs in Linux kernel capabilities code
	- Bug #2241: weights not applied properly in round-robin peer
	  selection
	- Off by one error in DNS label decompression could cause valid DNS
	  messages to be rejected
	- logformat docs contain extra whitespace
	- Reject ridiculously large ASN.1 lengths
	- Fix SNMP reporting of counters with a value > 0xFF80000
	- Correct spelling of WCCPv2 dst_port_hash to match the source
	- Plug some "squid -k reconfigure" memory leaks. Mostly SSL related.
	- Bug #1993: Memory leak in http_reply_access deny processing
	- Bug #2122: In some situations collapsed_forwarding could leak
	  private information
	- Bug #2376: Round-Robin becomes unbalanced when a peer dies and comes
	  back
	- Bug #2387: The calculation of the number of hash buckets need to
	  account for the memory size, not only disk size
	- Bug #2393: DNS requests retried indefinitely at full speed on failed
	  TCP connection
	- Bug #2393: DNS retransmit queue could get hold up
	- Correct socket syscalls statistics in commResetFD()

Changes to squid-2.6.STABLE20 (25 Apr 2008)

	- Bug #2263: Custom log formats fail to log file sizes >2GB properly
	  on 32-bit platforms
	- Fix stripping NT domain in squid_ldap_group
	- Bug #2278: Cache-Control: max-stale=0 forwarded wrongly as max-stale
	(without delta)
	- Bug #2283: Fails to parse chunked encoding using chunk extensions
	- Bug #420: Deal properly with empty list HTTP header members
	- Windows Server 2008 support
	- Bug #1886: tcp_outgoing_address acl doesn't work with indirect
	  source address (follow-x-forwarded-for)
	- Bug #2296: Stuck in 100% CPU when fetching an corrupt peer digest
	- Add support for the resolv.conf domain directive, and also
	  automatically derived default domain
	- minimum_icp_query_timeout directive
	- Bug #2329: Range header ignored on HIT

Changes to squid-2.6.STABLE19 (19 Mar 2008)

	- Fix tcp_outgoing_address example config to match its description
	- Bug #2198: assertion failed sc != NULL when using peer monitor
	  function
	- Fix missing default disk store type into QUICKSTART example.
	- Bugzilla #761 : Handle recursive completion operations in diskd.
	- documentation bugfix for tcp_outgoing_tos directive
	- Sort cache list in wccpv2 to ensure a consistent hash allocation
	  across all services
	- Updated Ukrainan error pages
	- Compile error in squid_kerb_auth under Mac OS X 10.5.2
	- squid_radius_auth failed ro process more than 256 requests
	- Clarified description of 'cache_vary' directive
	- Make range_offset_limit 0 disable local range processing as
	  documented, even if the first range starts at 0

Changes to squid-2.6.STABLE18 (10 Jan 2008)

	- Fix 2 assertion failures related to the fix for SQUID-2007:2
	- GPL license cleanup to GPLv2 or later. One file in edir_digest_auth
	  was GPLv2 only, now replaced with a GPLv2 or later licensed vesion.
	- Minor cleanups to make certain 64-bit platforms happier
	- Several Digest authentication bugs fixed wich was causing random
	  authenitcation popups or failures.
	- --with-valgrind-debug updated for valgrind-3.3.0.

Changes to squid-2.6.STABLE17 (26 Nov 2007)

	- Fix compile error with old GCC 2.x or other ANSI-C compilers before
	  C99
	- Mention the login= cache_peer option in release notes
	- Fix bad cache_peer example in squid.conf
	- Bug #2086: Fix a compile-time memory corruption error causing cf_gen
	  to fail
	- Bug #2048: Clarify high_memory_warning usage
	- Reject DNS responses which result in no data

helpers/basic_auth/LDAP/squid_ldap_auth.c
/*
 * squid_ldap_auth: authentication via ldap for squid proxy server
 * 
 * Maintainer: Henrik Nordstrom <hno@squid-cache.org>
 *
 * Author: Glen Newton 
 * glen.newton@nrc.ca
 * Advanced Services 
 * CISTI
 * National Research Council
 * 
 * Usage: squid_ldap_auth -b basedn [-s searchscope]
 *                        [-f searchfilter] [-D binddn -w bindpasswd]
 *                        [-u attr] [-h host] [-p port] [-P] [-R] [ldap_server_name[:port]] ...
 * 
 * Dependencies: You need to get the OpenLDAP libraries
 * from http://www.openldap.org
 * 
 * License: squid_ldap_auth is free software; you can redistribute it 
 * and/or modify it under the terms of the GNU General Public License 
 * as published by the Free Software Foundation; either version 2, 
 * or (at your option) any later version.
 *
 * Changes:
 * 2001-10-04: Henrik Nordstrom <hno@squid-cache.org>
 *             - Be consistent with the other helpers in how
 *               spaces are managed. If there is space characters
 *               then these are assumed to be part of the password
 * 2001-09-05: Henrik Nordstrom <hno@squid-cache.org>
 *             - Added ability to specify another default LDAP port to
 *               connect to. Persistent connections moved to -P
 * 2001-05-02: Henrik Nordstrom <hno@squid-cache.org>
 *             - Support newer OpenLDAP 2.x libraries using the
 *               revised Internet Draft API which unfortunately
 *               is not backwards compatible with RFC1823..
 * 2001-04-15: Henrik Nordstrom <hno@squid-cache.org>
 *             - Added command line option for basedn
 *             - Added the ability to search for the user DN
 * 2001-04-16: Henrik Nordstrom <hno@squid-cache.org>
 *             - Added -D binddn -w bindpasswd.
 * 2001-04-17: Henrik Nordstrom <hno@squid-cache.org>
 *             - Added -R to disable referrals
 *             - Added -a to control alias dereferencing
 * 2001-04-17: Henrik Nordstrom <hno@squid-cache.org>
 *             - Added -u, DN username attribute name
 * 2001-04-18: Henrik Nordstrom <hno@squid-cache.org>
 *             - Allow full filter specifications in -f
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <lber.h>
#include <ldap.h>

#include "util.h"

/* Change this to your search base */
static char *basedn;
static char *searchfilter = NULL;
static char *binddn = NULL;
static char *bindpasswd = NULL;
static char *userattr = "uid";
static int searchscope = LDAP_SCOPE_SUBTREE;
static int persistent = 0;
static int noreferrals = 0;
static int aliasderef = LDAP_DEREF_NEVER;

static int checkLDAP(LDAP * ld, char *userid, char *password);

/* Yuck.. we need to glue to different versions of the API */

#if defined(LDAP_API_VERSION) && LDAP_API_VERSION > 1823
static int
squid_ldap_errno(LDAP * ld)
{
    int err = 0;
    ldap_get_option(ld, LDAP_OPT_ERROR_NUMBER, &err);
    return err;
}
static void
squid_ldap_set_aliasderef(LDAP * ld, int deref)
{
    ldap_set_option(ld, LDAP_OPT_DEREF, &deref);
}
static void
squid_ldap_set_referrals(LDAP * ld, int referrals)
{
    int *value = referrals ? LDAP_OPT_ON : LDAP_OPT_OFF;
    ldap_set_option(ld, LDAP_OPT_REFERRALS, value);
}

#else
static int
squid_ldap_errno(LDAP * ld)
{
    return ld->ld_errno;
}
static void
squid_ldap_set_aliasderef(LDAP * ld, int deref)
{
    ld->ld_deref = deref;
}
static void
squid_ldap_set_referrals(LDAP * ld, int referrals)
{
    if (referrals)
	ld->ld_options |= ~LDAP_OPT_REFERRALS;
    else
	ld->ld_options &= ~LDAP_OPT_REFERRALS;
}
#endif

int
main(int argc, char **argv)
{
    char buf[256];
    char *user, *passwd;
    char *ldapServer = NULL;
    LDAP *ld = NULL;
    int tryagain;
    int port = LDAP_PORT;

    setbuf(stdout, NULL);

    while (argc > 1 && argv[1][0] == '-') {
	char *value = "";
	char option = argv[1][1];
	switch (option) {
	case 'P':
	case 'R':
	    break;
	default:
	    if (strlen(argv[1]) > 2 || argc <= 2) {
		value = argv[1] + 2;
	    } else {
		value = argv[2];
		argv++;
		argc--;
	    }
	    break;
	}
	argv++;
	argc--;
	switch (option) {
	case 'h':
	    if (ldapServer) {
		int len = strlen(ldapServer) + 1 + strlen(value) + 1;
		char *newhost = malloc(len);
		snprintf(newhost, len, "%s %s", ldapServer, value);
		free(ldapServer);
		ldapServer = newhost;
	    } else {
		ldapServer = strdup(value);
	    }
	    break;
	case 'b':
	    basedn = value;
	    break;
	case 'f':
	    searchfilter = value;
	    break;
	case 'u':
	    userattr = value;
	    break;
	case 's':
	    if (strcmp(value, "base") == 0)
		searchscope = LDAP_SCOPE_BASE;
	    else if (strcmp(value, "one") == 0)
		searchscope = LDAP_SCOPE_ONELEVEL;
	    else if (strcmp(value, "sub") == 0)
		searchscope = LDAP_SCOPE_SUBTREE;
	    else {
		fprintf(stderr, "squid_ldap_auth: ERROR: Unknown search scope '%s'\n", value);
		exit(1);
	    }
	    break;
	case 'a':
	    if (strcmp(value, "never") == 0)
		aliasderef = LDAP_DEREF_NEVER;
	    else if (strcmp(value, "always") == 0)
		aliasderef = LDAP_DEREF_ALWAYS;
	    else if (strcmp(value, "search") == 0)
		aliasderef = LDAP_DEREF_SEARCHING;
	    else if (strcmp(value, "find") == 0)
		aliasderef = LDAP_DEREF_FINDING;
	    else {
		fprintf(stderr, "squid_ldap_auth: ERROR: Unknown alias dereference method '%s'\n", value);
		exit(1);
	    }
	    break;
	case 'D':
	    binddn = value;
	    break;
	case 'w':
	    bindpasswd = value;
	    break;
	case 'P':
	    persistent = !persistent;
	    break;
	case 'p':
	    port = atoi(value);
	    break;
	case 'R':
	    noreferrals = !noreferrals;
	    break;
	default:
	    fprintf(stderr, "squid_ldap_auth: ERROR: Unknown command line option '%c'\n", option);
	    exit(1);
	}
    }

    while (argc > 1 && argv[1][0] == '-') {
	char *value = argv[1];
	if (ldapServer) {
	    int len = strlen(ldapServer) + 1 + strlen(value) + 1;
	    char *newhost = malloc(len);
	    snprintf(newhost, len, "%s %s", ldapServer, value);
	    free(ldapServer);
	    ldapServer = newhost;
	} else {
	    ldapServer = strdup(value);
	}
	argc--;
	argv++;
    }
    if (!ldapServer)
	ldapServer = "localhost";

    if (!basedn) {
	fprintf(stderr, "Usage: squid_ldap_auth -b basedn [options] [ldap_server_name[:port]]...\n\n");
	fprintf(stderr, "\t-b basedn (REQUIRED)\tbase dn under which to search\n");
	fprintf(stderr, "\t-f filter\t\tsearch filter to locate user DN\n");
	fprintf(stderr, "\t-u userattr\t\tusername DN attribute\n");
	fprintf(stderr, "\t-s base|one|sub\t\tsearch scope\n");
	fprintf(stderr, "\t-D binddn\t\tDN to bind as to perform searches\n");
	fprintf(stderr, "\t-w bindpasswd\t\tpassword for binddn\n");
	fprintf(stderr, "\t-h server\t\tLDAP server (defaults to localhost)\n");
	fprintf(stderr, "\t-p port\t\t\tLDAP server port\n");
	fprintf(stderr, "\t-P\t\t\tpersistent LDAP connection\n");
	fprintf(stderr, "\t-R\t\t\tdo not follow referrals\n");
	fprintf(stderr, "\t-a never|always|search|find\n\t\t\t\twhen to dereference aliases\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "\tIf no search filter is specified, then the dn <userattr>=user,basedn\n\twill be used (same as specifying a search filter of '<userattr>=',\n\tbut quicker as as there is no need to search for the user DN)\n\n");
	fprintf(stderr, "\tIf you need to bind as a user to perform searches then use the\n\t-D binddn -w bindpasswd options\n\n");
	exit(1);
    }
    while (fgets(buf, 256, stdin) != NULL) {
	user = strtok(buf, " \r\n");
	passwd = strtok(NULL, "\r\n");

	if (!user || !passwd || !passwd[0]) {
	    printf("ERR\n");
	    continue;
	}
	rfc1738_unescape(user);
	rfc1738_unescape(passwd);
	tryagain = 1;
      recover:
	if (ld == NULL) {
	    if ((ld = ldap_init(ldapServer, port)) == NULL) {
		fprintf(stderr, "\nUnable to connect to LDAP server:%s port:%d\n",
		    ldapServer, port);
		exit(1);
	    }
	    squid_ldap_set_referrals(ld, !noreferrals);
	    squid_ldap_set_aliasderef(ld, aliasderef);
	}
	if (checkLDAP(ld, user, passwd) != 0) {
	    if (tryagain && squid_ldap_errno(ld) != LDAP_INVALID_CREDENTIALS) {
		tryagain = 0;
		ldap_unbind(ld);
		ld = NULL;
		goto recover;
	    }
	    printf("ERR\n");
	} else {
	    printf("OK\n");
	}
	if (!persistent || (squid_ldap_errno(ld) != LDAP_SUCCESS && squid_ldap_errno(ld) != LDAP_INVALID_CREDENTIALS)) {
	    ldap_unbind(ld);
	    ld = NULL;
	}
    }
    if (ld)
	ldap_unbind(ld);
    return 0;
}

static int
checkLDAP(LDAP * ld, char *userid, char *password)
{
    char dn[256];

    if (!*password) {
	/* LDAP can't bind with a blank password. Seen as "anonymous"
	 * and always granted access
	 */
	return 1;
    }
    if (searchfilter) {
	char filter[256];
	LDAPMessage *res = NULL;
	LDAPMessage *entry;
	char *searchattr[] =
	{NULL};
	char *userdn;
	int rc;

	if (binddn) {
	    rc = ldap_simple_bind_s(ld, binddn, bindpasswd);
	    if (rc != LDAP_SUCCESS) {
		fprintf(stderr, "squid_ldap_auth: WARNING, could not bind to binddn '%s'\n", ldap_err2string(rc));
		return 1;
	    }
	}
	snprintf(filter, sizeof(filter), searchfilter, userid, userid, userid, userid, userid, userid, userid, userid, userid, userid, userid, userid, userid, userid, userid);
	if (ldap_search_s(ld, basedn, searchscope, filter, searchattr, 1, &res) != LDAP_SUCCESS) {
	    int rc = ldap_result2error(ld, res, 0);
	    if (noreferrals && rc == LDAP_PARTIAL_RESULTS) {
		/* Everything is fine. This is expected when referrals
		 * are disabled.
		 */
	    } else {
		fprintf(stderr, "squid_ldap_auth: WARNING, LDAP search error '%s'\n", ldap_err2string(rc));
	    }
	}
	entry = ldap_first_entry(ld, res);
	if (!entry) {
	    ldap_msgfree(res);
	    return 1;
	}
	userdn = ldap_get_dn(ld, entry);
	if (!userdn) {
	    fprintf(stderr, "squid_ldap_auth: ERROR, could not get user DN for '%s'\n", userid);
	    ldap_msgfree(res);
	    return 1;
	}
	snprintf(dn, sizeof(dn), "%s", userdn);
	free(userdn);
	ldap_msgfree(res);
    } else {
	snprintf(dn, sizeof(dn), "%s=%s,%s", userattr, userid, basedn);
    }

    if (ldap_simple_bind_s(ld, dn, password) != LDAP_SUCCESS)
	return 1;

    return 0;
}
